Sources/SwiftProtobuf/StringUtils.swift

```diff
@@ -8,32 +8,115 @@
 ///
-/// Utility functions that are generally useful...
+/// Utility functions for converting UTF8 bytes into Strings.
+/// These functions must:
+///  * Accept any valid UTF8, including a zero byte (which is
+///    a valid UTF8 encoding of U+0000)
+///  * Return nil for any invalid UTF8
+///  * Be fast (since they're extensively used by all decoders
+///    and even some of the encoders)
 ///
 // -----------------------------------------------------------------------------
 
 import Foundation
 
-// Convert a pointer/count that refers to a block of UTF8 bytes
-// in memory into a String.
-// Returns nil if UTF8 is invalid.
+// Wrapper that takes a buffer and start/end offsets
+internal func utf8ToString(
+  bytes: UnsafeBufferPointer<UInt8>,
+  start: UnsafeBufferPointer<UInt8>.Index,
+  end: UnsafeBufferPointer<UInt8>.Index
+) -> String? {
+  return utf8ToString(bytes: bytes.baseAddress! + start, count: end - start)
+}
+
+// Swift's support for working with UTF8 bytes directly has
+// evolved over time.  The following tries to choose the
+// best option depending on the version of Swift you're using.
+
+#if swift(>=4.0)
+
+///////////////////////////
+//
+// Swift 4 (all platforms)
+//
+////////////////////////////
+
+// Swift 4 introduced new faster String facilities
+// that seem to work consistently across all platforms.
+
+internal func utf8ToString(bytes: UnsafePointer<UInt8>, count: Int) -> String? {
+    if count == 0 {
+        return String()
+    }
+    let codeUnits = UnsafeBufferPointer<UInt8>(start: bytes, count: count)
+    let sourceEncoding = Unicode.UTF8.self
+
+    // Verify that the UTF-8 is valid.
+    var p = sourceEncoding.ForwardParser()
+    var i = codeUnits.makeIterator()
+  Loop:
+    while true {
+      switch p.parseScalar(from: &i) {
+      case .valid(_):
+        break
+      case .error:
+        return nil
+      case .emptyInput:
+        break Loop
+      }
+    }
+
+    // This initializer is fast but does not reject broken
+    // UTF-8 (which is why we validate the UTF-8 above).
+    return String(decoding: codeUnits, as: sourceEncoding)
+ }
+
+#elseif os(OSX) || os(tvOS) || os(watchOS) || os(iOS)
+
+//////////////////////////////////
+//
+// Swift 3 (Apple platforms)
+//
+//////////////////////////////////
 
-#if !os(Linux) || swift(>=3.1)
-fileprivate func fastUtf8ToString(bytes: UnsafePointer<UInt8>, count: Int) -> String? {
+internal func utf8ToString(bytes: UnsafePointer<UInt8>, count: Int) -> String? {
+    if count == 0 {
+        return String()
+    }
+    // On Apple platforms, the Swift 3 version of Foundation has a String
+    // initializer that works for us:
     let s = NSString(bytes: bytes, length: count, encoding: String.Encoding.utf8.rawValue)
     if let s = s {
         return String._unconditionallyBridgeFromObjectiveC(s)
     }
     return nil
 }
-#endif
 
-#if !os(Linux) || swift(>=3.2)
-// Deliberately empty.  macOS and Linux >=3.2 don't need the slow version.
 #else
```

Comment:
Do you want to use elseif os(linux) and let the build break if we ever get an unexpected platform for < swift 4?